### ISSUE TITLE
dssm实现方式

### ISSUE THREAD

**Issue by windy444** [PaddlePaddle/Paddle#5297]
看了下paddle提供的dssm实现代码，最上层是用的square_error_cost，这个和论文里面的最后一层区别还是很大的。请问paddle有按照dssm论文里面的实现方式吗？

**Comment by wangkuiyi** [comment]
可否贴一下你看到的DSSM代码的 Github permlink吗？以及论文的链接？这样方便找到对应模块的负责人，设置为这个issue的 assignee。

获取 Github permlink 的方式如下：https://help.github.com/articles/getting-permanent-links-to-files/

**Comment by cxysteven** [comment]
@windy444 如果你说的论文是“Learning deep structured semantic models for web search using clickthrough data“.   论文里面是cosine后接softmax，你可以把square_error_cost改成相应的classification_cost（前提是离散标签).
如果是想进行直接优化论文中NDCG指标
可以参考：https://github.com/PaddlePaddle/models/tree/develop/ltr

**Comment by lcy-seso** [comment]
DSSM 作为一种框架，有若干相关的论文和变种。Paddle Models 中的DSSM有若干流程分支，我们按照 DSSM 的框架和思想来实现，并对模型验证过。

是否可以提供您参考的论文？以及是否在使用DSSM模型中遇到什么问题呢？

**Comment by windy444** [comment]
我说的论文是：Learning deep structured semantic models for web search using clickthrough data
我这边的训练数据标签就是0，1两种，点和不点，所以不清楚这种情况下square_error_cost能不能达到好的效果。
cosine后接softmax，论文里面应该是把左右两边的向量做类似softmax的计算，而标准的softmax应该是得到一个向量后，和参数矩阵做计算。我理解这个里面是不太一样的吧

**Comment by lcy-seso** [comment]
Models 中的DSSM 模型通过指定`model_type` （分类0，排序1，回归2）可以在：softmax + cross entropy、rank loss，和 square error 三种损失函数之间切换。

哪一种更合适具体的问题/数据，通常也需要一些简单的实验。

**Comment by windy444** [comment]
@lcy-seso paddle dssm里面实现的分类问题，是把左右两个向量concat起来，传入softmaxt。如果对于求解item对应的向量这种问题，我不太确定，左右两个向量是否能真正有意义。

**Comment by lcy-seso** [comment]
我看了您提到的论文，Models中的DSSM使用的优化准则和您提到的这篇论文中的优化准则不同。

论文中定义的这种损失函数从策略上接近于 LTR 中的 “Listwise” 策略。用同query下所有检索到候选来单个候选进行归一化，最后进行极大似然估计作为优化准则。

而 DSSM 目前默认的 softmax/regression 都是 "Pointwise"，不考虑到query检索出的其它候选。Paddle没有直接对应于论文中的损失函数（sequence softmax 可以基于一个list进行归一化，但没有损失函数能够利用softmax的计算结果）。

如果学习目标是得到item的表示，损失函数应该确实会起到举足轻重的作用。是否能够以其它损失函数得到同样的学习效果，甚至是更好的学习效果，我也不十分肯定，机器学习任务也都需要一些实验。

**Comment by windy444** [comment]
@lcy-seso 感谢回复。期待paddle可以实现下论文中的方式。我这边暂时考虑youtube推荐论文的方式了，这个paddle应该支持的。

**Comment by lcy-seso** [comment]
我个人觉得论文中这个损失函数最核心的部分应该是能够考虑整个List的信息。比较接近 LTR 中的Listwise策略。Paddle中目前只有LambdaRank是Listwise类型的损失函数。但DSSM与LTR追求“序”还是有所不同，DSSM 追求特征空间的相似性度量。LambdaRank 对学习相似性表示，从原理上不像是一个特别好的选择。

目前没有现成和论文中一模一样可用的，我觉得可以用 Pairwise 策略来近似看看学习效果（如果可以接受的话）。

**Comment by lcy-seso** [comment]
确实，损失函数是学习任务的核心之一 ~  支持各种优化考量应该确实挺重要的。我们会考虑的~ 谢谢啦~
